CMakeLists.txt

```diff
@@ -0,0 +1,103 @@
+# set range of known cmake versions where all policies are fine for us
+cmake_minimum_required(VERSION 3.1...3.13)
+if(${CMAKE_VERSION} VERSION_LESS 3.12) # do the best we can in this case
+    cmake_policy(VERSION ${CMAKE_MAJOR_VERSION}.${CMAKE_MINOR_VERSION})
+endif()
+
+file(STRINGS VERSION VERSION) # read hlwm version from file 'VERSION'
+project(Herbstluftwm
+	VERSION ${VERSION}
+	DESCRIPTION "A manual tiling window manager for X11")
+
+# write all executables to root of build directory
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
+
+## macros & functions (may be moved to cmake/ and included)
+# get project dependencies
+macro (get_external_deps) # – macro for scope
+	find_package(X11 REQUIRED)
+	include(FindPkgConfig)
+	pkg_check_modules(GLIB2 REQUIRED glib-2.0)
+endmacro()
+
+# extract git commit hash and append to project version
+function (set_gitversion)
+	set(PROJECT_VERSION_GIT ${PROJECT_VERSION} PARENT_SCOPE)
+	if (EXISTS ${CMAKE_SOURCE_DIR}/.git)
+		find_package(Git)
+		if (GIT_FOUND)
+			execute_process(
+				COMMAND git rev-parse --short HEAD
+				WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+				OUTPUT_VARIABLE GIT_HASH
+				OUTPUT_STRIP_TRAILING_WHITESPACE
+			)
+			string(TIMESTAMP DATE "%Y-%m-%d")
+			set(PROJECT_VERSION_GIT 
+				"${PROJECT_VERSION} (${GIT_HASH} built on ${DATE})"
+				PARENT_SCOPE)
+		endif()
+	endif()
+endfunction()
+
+# export version string as a define
+function (export_version SOURCEFILE)
+	set_property(SOURCE ${SOURCEFILE} APPEND PROPERTY COMPILE_DEFINITIONS
+		HERBSTLUFT_VERSION=\"${PROJECT_VERSION_GIT}\")
+endfunction()
+
+# add custom build flags to the DEBUG build type
+function (add_debugflags TARGET)
+	set(FLAGS
+		-Og
+		-pedantic -Wall -Werror
+		-Wno-sign-compare -Wno-narrowing -Wno-error=cast-function-type
+		-fsanitize=address,leak,undefined
+		-fno-omit-frame-pointer -fuse-ld=gold)
```

Comment:
I took gold linker option from travis. But I am not so sure if we should put that here and if this option is supported by clang. What's the rationale?

Reply:
I also don't know which option is for which compiler. @dnnr?

Reply:
IIRC, I had to use gold when using gcc-8.2 with sanitizers, due to some bug in GCC (missing space between arguments passed to the linker). Generally, gold is better™, but there is no inherent need for it. So if you get it work without it, it's fine.